CSharp/Library/QnAMaker/QnAMakerDialog.cs

```diff
@@ -1,110 +1,225 @@
+        
+        public IQnAService[] MakeServicesFromAttributes()
+        {
+            var type = this.GetType();
+            var qnaModels = type.GetCustomAttributes<QnAMakerAttribute>(inherit: true);
+            return qnaModels.Select(m => new QnAMakerService(m)).Cast<IQnAService>().ToArray();
+        }
+
+        /// <summary>
+        /// Construct the QnA Service dialog.
+        /// </summary>
+        /// <param name="services">The QnA service.</param>
+        public QnAMakerDialog(params IQnAService[] services)
+        {
+            if (services.Length == 0)
+            {
+                services = MakeServicesFromAttributes();
+            }
+
+            SetField.NotNull(out this.services, nameof(services), services);
+        }
+
+        async Task IDialog<IMessageActivity>.StartAsync(IDialogContext context)
+        {
+            context.Wait(MessageReceivedAsync);
+        }
+
+        public async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> argument)
+        {
+            var message = await argument;
+            var qnaMakerResults = default(QnAMakerResults);
+
+            if (message != null && !string.IsNullOrEmpty(message.Text))
+            {
+                var tasks = this.services.Select(s => s.QueryServiceAsync(message.Text)).ToArray();
+
+                if (tasks.Any())
+                {
+                    var maxValue = tasks.Max(x => x.Result.Answers[0].Score);
+                    qnaMakerResults = tasks.First(x => x.Result.Answers[0].Score == maxValue).Result;
+
+                    var filteredQnAResults = new List<QnAMakerResult>();
+                    foreach (var qnaMakerResult in qnaMakerResults.Answers)
+                    {
+                        qnaMakerResult.Score /= 100;
+                        if (qnaMakerResult.Score >= qnaMakerResults.ServiceCfg.ScoreThreshold)
+                        {
+                            qnaMakerResult.Answer = HttpUtility.HtmlDecode(qnaMakerResult.Answer);
+                            filteredQnAResults.Add((qnaMakerResult));
+                        }
+                    }
+
+                    qnaMakerResults.Answers = filteredQnAResults;
+
+                    if (qnaMakerResults != null && qnaMakerResults.Answers != null && qnaMakerResults.Answers.Count > 0)
+                    {
+                        if (this.IsConfidentAnswer(qnaMakerResults))
+                        {
+                            await this.RespondFromQnAMakerResultAsync(context, message, qnaMakerResults);
+                            await this.DefaultWaitNextMessageAsync(context, message, qnaMakerResults);
+                        }
+                        else
+                        {
+                            var feedbackRecord = new FeedbackRecord { UserId = message.From.Id, UserQuestion = message.Text };
+                            context.PrivateConversationData.SetValue("qnaMakerResults", qnaMakerResults);

```

Comment:
since there are maked as Serializable you can keep them as local members of dialog.